proof/crefine/AARCH64/Syscall_C.thy

```diff
@@ -2014,13 +2030,14 @@ proof -
        apply (rule_tac Q'="\<lambda>_ s. ?PRE s \<and> armHSCurVCPU (ksArchState s) = Some (vcpuPtr, active)" in hoare_post_imp)
         apply clarsimp
     subgoal for _ _ eisr0 eisr1
-      apply (clarsimp simp: invs'_HScurVCPU_vcpu_at' valid_arch_state'_def max_armKSGICVCPUNumListRegs_def dest!: invs_arch_state')
-      using sint_ctz[where x=eisr0, simplified]
-      apply (clarsimp simp: word_sless_alt word_sle_eq sint_int_ctz_ge_0 split: if_split)
-      using sint_ctz[where x=eisr1, simplified]
-      apply (subst signed_arith_sint; clarsimp simp: word_size; simp add: sint_int_ctz_ge_0)
-      using sint_ctz[where x=eisr1, simplified]
-      apply (subst signed_arith_sint; clarsimp simp: word_size; simp add: sint_int_ctz_less_32)
+      apply (clarsimp simp: invs'_HScurVCPU_vcpu_at' valid_arch_state'_def dest!: invs_arch_state')
+      apply (clarsimp simp: eisr_calc_def not_le word_sless_alt word_sle_eq)
+      (* fold max_armKSGICVCPUNumListRegs_int_val to prevent arithmetic with the "+ 32" term *)
```

Comment:
here's a thought: is there any benefit to changing `GIC_VCPU_MAX_NUM_LR` in C to use an enum? it's a number small enough to fit into an int, would prevent this situation, but would probably break a casting or two. then you could have a lemma that `Kernel_C.appropriately_renamed_GIC_VCPU_MAX_NUM_LR = max_armKSGICVCPUNumListRegs_int_val` somewhere?

Reply:
In general yes, but I don't think it's needed for this one. The number is not freely configurable and it's not used in many places. It can only have exactly one of two values and they occur in controlled circumstances in the C. I.e. the space for testing this is small and we already cover it just by running the proofs for all platforms. If it ever goes out of sync, we'll get failures immediately. That's different to NUM_DOMAINS or others where the user can provide a number and we can't test the entire space.

Reply:
You're right that it would get rid of the `_val` numeral theorems. I still have the urge to keep those C changes to a minimum, and reserve annoyance for the kernel side for those where we really do need it.

Reply:
It's a strategic question; do we push constants to be constants in C over time, given there is a problem with defining constants in C. I guess you're right in this case, and it's only one of two values, but you understand that the general principle of folding a `#define` that appears in C might fail based on it getting mutated by simp, which is part of cinit, etc. In each specific case we can argue whether it's passable or not (e.g. in this case it is), but then if we have to test them in aggregate, I wonder where that leads. Fine to leave for this one, but do keep a vision of the future in mind.